clientd3d/d3drender.c

```diff
@@ -5197,6 +5274,255 @@ void D3DRenderFontInit(font_3d *pFont, HFONT hFont)
+			return;
+
+		d3d_render_chunk_new* pChunk = D3DRenderChunkNew(pPacket);
+		assert(pChunk);
+
+		pPacket->pMaterialFctn = &D3DMaterialWorldPacket;
+
+		pChunk->numIndices = 4;
+		pChunk->numVertices = 4;
+		pChunk->numPrimitives = pChunk->numVertices - 2;
+		pChunk->pMaterialFctn = &D3DMaterialWorldDynamicChunk;
+		pChunk->flags |= D3DRENDER_NOAMBIENT | D3DRENDER_WORLD_OBJ;
+		pChunk->zBias = ZBIAS_BASE;
+
+		int piAngle = overlay->x;
+		int piHeight = overlay->y;
+
+		auto mapRange = [](double value, double in_min, double in_max, double out_min, double out_max) -> double {
+			return (value - in_min) * (out_max - out_min) / (in_max - in_min) + out_min;
+		};
+
+		long mappedHeightValue = mapRange(piHeight, -200, 200, heightMin, heightMax);
+		float angleInRadians = piAngle / 4096.0f;
+		double azimuthalAngle = angleInRadians * 2 * PI;
+		long radius = heightMax * 5;
+		double horizontalRadius = sqrt(pow(radius, 2) - pow(mappedHeightValue, 2));
+
+		long x = params->viewer_x - (object_width / 2) + horizontalRadius * cos(azimuthalAngle);
+		long y = params->viewer_y - (object_height / 2) + horizontalRadius * sin(azimuthalAngle);
+		long z = params->viewer_height + mappedHeightValue;
+
+		Vector3D bg_overlay_pos;
+		bg_overlay_pos.x = x;
+		bg_overlay_pos.y = y;
+		bg_overlay_pos.z = z;
+
+		D3DMATRIX rot, mat;
+		MatrixIdentity(&mat);
+		MatrixIdentity(&rot);
+
+		MatrixRotateY(&rot, (float)angleHeading * 360.0f / 4096.0f * PI / 180.0f);
+		MatrixRotateX(&mat, (float)anglePitch * 50.0f / 414.0f * PI / 180.0f);
+		MatrixTranspose(&rot, &rot);
+		MatrixTranslate(&mat, bg_overlay_pos.x, bg_overlay_pos.z, bg_overlay_pos.y);
+		MatrixMultiply(&pChunk->xForm, &rot, &mat);
+
+		pChunk->xyz[0].x = 0;
+		pChunk->xyz[0].z = 0;
+		pChunk->xyz[0].y = 0;
+
+		pChunk->xyz[1].x = -object_width;
+		pChunk->xyz[1].z = 0;
+		pChunk->xyz[1].y = 0;
+
+		pChunk->xyz[2].x = -object_width;
+		pChunk->xyz[2].z = object_height;
+		pChunk->xyz[2].y = 0;
+
+		pChunk->xyz[3].x = 0;
+		pChunk->xyz[3].z = object_height;
+		pChunk->xyz[3].y = 0;
+
+		for (int j = 0; j < 4; j++) {
+			pChunk->bgra[j].b = 255;
+			pChunk->bgra[j].g = 255;
+			pChunk->bgra[j].r = 255;
+			pChunk->bgra[j].a = 255;
+		}
+
+		pChunk->st0[0].s = 0.001f;
+		pChunk->st0[0].t = 0.001f;
+		pChunk->st0[1].s = 0.001f;
+		pChunk->st0[1].t = 0.999f;
+		pChunk->st0[2].s = 0.999f;
+		pChunk->st0[2].t = 0.999f;
+		pChunk->st0[3].s = 0.999f;
+		pChunk->st0[3].t = 0.001f;
+
+		pChunk->indices[0] = 1;
+		pChunk->indices[1] = 2;
+		pChunk->indices[2] = 0;
+		pChunk->indices[3] = 3;
+
+		// Determine if the background overlay is visible for click detection in client
+		// (e.g. Blinded by the light of the ++Sun).
+		D3DMATRIX	localToScreen, trans;
+		custom_xyzw	topLeft, topRight, bottomLeft, bottomRight, center;
+		ObjectRange* range = FindVisibleObjectById(overlay->obj.id);
+		int			w, h;
+		int			tempLeft, tempRight, tempTop, tempBottom;
+		int			distX, distY, distance;
+
+		if (overlay->obj.id == player.id)

```

Comment:
I don't understand this check... we think the player might be a background overlay?